### DISCUSSION TITLE
[LLVM] interoperability with  std:vector: LLVMFunction _Znwm cannot be found

### DISCUSSION THREAD

**Discussion by bademux** [oracle/graal#6491] [Q&A] [answered]
Hello,
I have a problem with `std::vector` allocation.  Got error on line:
```C ret->items = new Items();```

```java
org.graalvm.polyglot.PolyglotException: External LLVMFunction _Znwm cannot be found.
	at <llvm>.allocNativeProgram(lib.c:25)
	at app/org.graalvm.sdk/org.graalvm.polyglot.Value.execute(Value.java:878)
```

Full listing
```C
#include <stdio.h>
#include <vector>
#include <graalvm/llvm/polyglot.h>

typedef struct Item {
    int i;
} Item;

typedef std::vector<Item> Items;

typedef struct Program {
    Items* items;
} Program;

POLYGLOT_DECLARE_STRUCT(Program)

extern "C" void *allocNativeProgram() {
  struct Program *ret = (Program *) malloc(sizeof(Program));
  ret->items = new Items();
  return polyglot_from_Program(ret);
}
```

compiled with:
```bash
clang -x c++ src/main/c/lib.c -emit-llvm -c -o src/test/resources/lib.bc
```

Thanks!

**Comment by rschatz** [comment, marked as answer]
This is using the wrong C++ standard library. It's compiling for `libstdc++`, you need to compile for `libc++` instead. Also, this does not link against the C++ library, so Sulong is not going to load it.

Try compiling your bitcode with the LLVM toolchain wrappers provided by Sulong, instead of vanilla clang. You can get the path to them with `lli --print-toolchain-path`. Also, don't use `-emit-llvm`. The toolchain wrappers will produce standard ELF files with embedded bitcode. That way, the dependency to the C++ library is visible to Sulong.

**Reply by bademux** [reply]
Great it helps! 
the problem solved by using `clang++` instead of `clang`. 
It somehow links to the `libc++` (not sure about this) 

`clang++ -xc++ -shared src/main/c/lib.c -o build/resources/main/lib.so`
```bash
$ clang++ --version
clang version 14.0.6 (GraalVM.org llvmorg-14.0.6-3-gc7a4a53c32-bgc5e298fd27 c7a4a53c32ccf93dd118fb6c13f2ecf02baeadf4)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/user/.sdkman/candidates/java/22.3.r19-grl/lib/llvm/bin
```

**Reply by rschatz** [reply]
>It somehow links to the libc++ (not sure about this)

If you want to know details, you can call it with `-v`. Then it will show you exactly what it's doing, what flags it's injecting and what linker command it is running.
